src/app/chat/chat.component.ts

```diff
@@ -27,34 +27,60 @@ export class ChatComponent {
     userId: string | null;
     username: string | null;
   }> = this.store.select(selectUserData);
-  protected readonly datasource: IDatasource<Message> = this.getDatasource();
+
+  protected readonly datasource = new Datasource<Message>({
+    get: (index: number, count: number) => {
+      console.log(index, count);
+      return this.store
+        .select(selectMessagesSlice(index, count))
+        .pipe(take(1), tap(console.log));
+    },
+    devSettings: { debug: true },
+  });
 
   constructor(
     private fb: FormBuilder,
     private store: Store<EntityState<Message>>
   ) {
+    let prev: Message[] = [];
     this.messages$
       .pipe(
-        tap(messages =>
-          this.datasource.adapter?.reset(this.getDatasource(messages.length))
-        )
+        tap(async curr => {
+          const newItem = curr.find(c => !prev.find(p => c.id == p.id));
+          if (newItem) {
+            prev = curr;
+            this.appendMessage(newItem);
+            return;
+          }
+          const oldItem = curr.find(c =>
+            prev.find(p => c.id == p.id && c.time !== p.time)
+          );
+          if (oldItem) {
+            prev = curr;
+            this.replaceMessage(oldItem);
+            return;
+          }
+        })
```

Comment:
Here's the core change. By prev/curr arrays we can detect change and make a decision: add new item or update the existed one. New item is detected by new id. Item to update is detected by time change. I had to reconsider the id logic at the effects level, see below.

Reply:
interesting. I would love to avoid nested iterations though, if possible.
I'm thinking I could trigger new or old message updates at the store effects, calling a service that holds the datasource.

Reply:
The Scroller doesn't know if there are any changes in the data before it requests this data via datasource.get method. So if the changing items are present in the DOM (say, not virtual), we need to think how to tell the Scroller about it. And the adapter methods are designed to provide such changes in the runtime.